Add SHA-256 and SHA-512-256 as authentication digest algorithms

gtjoseph opened the pull request:
* Refactored pjproject code to support the new algorithms and
added a patch file to third-party/pjproject/patches

* Added new parameters to the pjsip auth object:
  * password_digest = `<algorithm>:<digest>`
  * supported_algorithms_uac = List of algorithms to support
    when acting as a UAC.
  * supported_algorithms_uas = List of algorithms to support
    when acting as a UAS.
  See the auth object in pjsip.conf.sample for detailed info.

* Updated both res_pjsip_authenticator_digest.c (for UAS) and
res_pjsip_outbound_authentocator_digest.c (UAC) to suport the
new algorithms.

The new algorithms are only available with the bundled version
of pjproject, or an external version > 2.14.1.  OpenSSL version
1.1.1 or greater is required to support SHA-512-256.

Resolves: #948

UserNote: The SHA-256 and SHA-512-256 algorithms are now available
for authentication as both a UAS and a UAC.

gtjoseph commented:
This PR is in draft status because the pjproject patch has not been merged upstream yet.

gtjoseph commented:
cherry-pick-to: 20
cherry-pick-to: 21
cherry-pick-to: 22

gtjoseph commented:
Just pushed an update to res_pjsip_authenticator_digest:digest_check_auth that skips trying to verify artificial auths since they can only be used to send challenges and can never be verified.

gtjoseph commented:
pjproject has approved the pull request related to the new patch in third-party/pjproject/patches and marked it for release with pjproject-2.15 therefore I'm marking this PR as ready-to-review.

gtjoseph commented:
pjproject patch merged

silentindark commented on third-party/pjproject/patches/0001-Add-full-support-for-SHA-256-and-SHA-512-256-digest-.patch:
```diff
@@ -0,0 +1,2443 @@
++        case PJSIP_CRED_DATA_PLAIN_PASSWD:
++            /* PLAIN_PASSWD creds can be used for any algorithm other than AKA */
++            if (algorithm_type != PJSIP_AUTH_ALGORITHM_AKAV1_MD5
++                    || algorithm_type != PJSIP_AUTH_ALGORITHM_AKAV2_MD5) {
```
Small change in this commit: https://github.com/pjsip/pjproject/commit/c28635a2d3d27317c9965f7b9182063f20ca5a28.patch
From:
if (algorithm_type != PJSIP_AUTH_ALGORITHM_AKAV1_MD5
+                    || algorithm_type != PJSIP_AUTH_ALGORITHM_AKAV2_MD5) {
To:
            if (algorithm_type != PJSIP_AUTH_ALGORITHM_AKAV1_MD5
                    && algorithm_type != PJSIP_AUTH_ALGORITHM_AKAV2_MD5) {

gtjoseph replied:
@silentindark Thanks for catching this.  Teluu will be releasing pjproject 2.15 shortly so we'll wait until we update bundled-pjproject to 2.15.  That way we'll get the new stuff plus the fix without needing the patch in third-party/pjproject/patches.

Moving back to draft status.

gtjoseph replied:
Unresolving so the comment stays visible.

jcolp commented on res/res_pjsip/config_auth.c:
```diff
@@ -24,13 +24,101 @@
+ * NOTE: We don't support AKA but we need to specify
+ * it to be compatible with the pjproject definition.
+ */
+#include "openssl/md5.h"
```
We can be built without OpenSSL, so in that case I think this would cause a build failure.

gtjoseph replied:
done

jcolp commented on res/res_pjsip_authenticator_digest.c:
```diff
@@ -131,58 +139,132 @@ static const struct ast_sip_auth *get_auth(void)
 	if (auth->type == AST_SIP_AUTH_TYPE_ARTIFICIAL) {
-		return PJSIP_SC_FORBIDDEN;
+		/*
+		 * This shouldn't happen because this function can only be invoked
```
Wouldn't this be the case if we challenge on artificial auth, and they resent with Authorization?

gtjoseph replied:
There's no scenario where we would actually verify an artificial auth.  We can challenge with an artificial auth if the endpoint only has auth_username for instance, but when the challenge comes back it has a userid so we look that up to find the endpoint and auth object.  Besides, the artificial auth has no credentials to verify against.  This is checked down in digest_check_auth().

jcolp commented:
Now needs #1020 before merging.

gtjoseph commented:
NOTE:  The only changes to this PR since last approval have been...
* Rebased to pull in bundled pjproject 2.15.1
* Updated configs/samples/pjsip.conf.sample
* Updated res/res_pjsip/pjsip_config.xml

You only need to re-review the two updated files.
